src/vm/functions/database.rs

```diff
@@ -38,6 +38,50 @@ pub fn special_contract_call(args: &[SymbolicExpression],
         contract_name, function_name, &rest_args)
 }
 
+pub fn special_fetch_variable(args: &[SymbolicExpression],
+                              env: &mut Environment,
+                              context: &LocalContext) -> Result<Value> {
+    // arg0 -> var name
+    if args.len() != 1 {
+        return Err(UncheckedError::InvalidArguments("(fetch-entry ...) requires exactly 1 argument".to_string()).into())
+    }
+
+    let var_name = args[0].match_atom()
+        .ok_or(UncheckedError::InvalidArguments("First argument in fetch-var must be the var name".to_string()))?;
+
+    let value = env.global_context.database.lookup_variable(&env.contract_context.name, var_name)?;
+    match value {
+        Some(data) => Ok(data),
+        None => Ok(Value::none())
+    }
+}
+
+pub fn special_set_variable(args: &[SymbolicExpression],
+                            env: &mut Environment,
+                            context: &LocalContext) -> Result<Value> {
+    if env.global_context.is_read_only() {
+        return Err(UncheckedError::WriteFromReadOnlyContext.into())
+    }
+
+    // arg0 -> var name
+    // arg1 -> value
+    if args.len() != 2 {
+        return Err(UncheckedError::InvalidArguments("(set-var! ...) requires exactly 2 arguments".to_string()).into())
+    }
+
+    let value = eval(&args[1], env, &context)?;
+
+    let var_name = args[0].match_atom()
+        .ok_or(UncheckedError::InvalidArguments("First argument in set-var! function must be the map name".to_string()))?;
+
+    // (define-data-var my-var int) is implicitely implying that my-var is optional.
+    // (setvar! my-var 1) however, will return an Int type signature. 
+    // As such, we will wrap `value` in an Optional<Int>. 
+    let optional_value = Value::Optional(OptionalData { data: Some(Box::new(value)) });
```

Comment:
I think the interactions here and in `lookup_variable` will be slightly cleaner if you don't wrap this in an optional here, but instead do it on lookup.

That means, you'd delete `let optional_value = ...` here, and instead store the `value` directly.
`database.lookup_variable` would return `Ok(None)` values instead of `Ok(Some(NONE))`, and then `special_fetch_variable` would just always wrap the `Value::Optional`.

Reply:
Addressed with https://github.com/blockstack/blockstack-core/pull/1022/commits/a105805bb6382d6b841c7730ad3673cd3c08d851